nixos/modules/services/continuous-integration/gitlab-runner.nix

```diff
@@ -35,12 +99,20 @@ let
       gitlab-runner verify --delete
 
       # current and desired state
-      NEEDED_SERVICES=$(echo ${concatStringsSep " " (attrNames hashedServices)} | tr " " "\n")
+      CURRENT_SERVICES=$(cat $REGISTERED_STATE_FILE || true)
+      DESIRED_SERVICES=$(echo ${escapeShellArg (concatStringsSep "\n" hashedServices)})
       REGISTERED_SERVICES=$(gitlab-runner list 2>&1 | grep 'Executor' | awk '{ print $1 }')
 
       # difference between current and desired state
-      NEW_SERVICES=$(grep -vxF -f <(echo "$REGISTERED_SERVICES") <(echo "$NEEDED_SERVICES") || true)
-      OLD_SERVICES=$(grep -vxF -f <(echo "$NEEDED_SERVICES") <(echo "$REGISTERED_SERVICES") || true)
+      NEW_SERVICES=$(${getRunnerServices} register --current "$CURRENT_SERVICES" --desired "$DESIRED_SERVICES" --registered "$REGISTERED_SERVICES")
```

Comment:
I find the semantic awkward, it's the same call to register than for unregister and we still have a shell script that calls gitlab-runner unregister, what's the point of calling `${getRunnerServices} unregister` then ?

Reply:
`${getRunnerServices} unregister` outputs a list of services which needs to be unregistered.